### DISCUSSION TITLE
Allow osu! tournaments client act as a referee in multiplayer rooms

### DISCUSSION THREAD

**Discussion by astro-angelfish** [ppy/osu#29008] [Ideas]
I'm developing on an osu! bot framework and aiming to create an automated osu! tournament referee because I feel like it could be really automated. And I think the programmatic functions such as picking maps in pool and scoring could just get merged in osu!tournament client as they have the same permission in the tournament room. It could free carbon-based referees from waiting for their players and getting bored. It also could reduce the mistakes that carbon-based referee would make. #25599 is talking about lack of tournament staffs so I think it is a possible but partial solution to that.

Therefore, a pause function is also needed because there might be incidents such as disconnection, forfeiting, account restriction and etc that carbon-based referee must play in. Also uncommon rules used by tournaments might also need carbon-based referee or just programmed as a client plugin. So parts of its map picking functionality might also need also to be able to be disabled.

<details><summary>Current details I've come up with.</summary>

- Special mod score configuration file might be required for `FM` maps. Such as `1.8x0.5=0.9`x score on EZ mod. It is also a good idea to allow score multiplier temporary modified in the multiplayer room according to configuration. But it would increase memory load of the server and lead to large modification to osu. The mod score override configuration could be easily covered by `.ini` files.
- Carbon-based referee should always have the highest permission to control the multiplayer room. The tournament client is just an assistant.
- Players are expected to use command such as `!pick <id>` to pick maps.
- Players should be per-configured before the room creation. After the creation the players in list are expected to get invited immediately.
- Carbon-based referee should be able to modify score counted by the client. Both in GUI and command.
- Tournament client need to accept a pause request from players. And the pause persists until carbon-based referees decide to resume the tournament.
  - Tournament holders are expected to set rules about pausing and punishments about abusing.
  - If not all players are ready, tournament client can choose to activate a pause routine automatically or start anyway base on its configuration.
- Tournament client should automatically and forcefully pick map with id `TB` if both team/player reached their final score to win. But if there is multiple `TB` maps or no maps with id `TB`, carbon-based referee should come in.
  - Client should not respond to pick request from player anymore.
- A plugin system is optional. It is used to adapt the customized rules for each tournaments. But holders should be aware to malwares.
- Player not in participate list joined room should be immediately kicked.
- All of functions could be implemented by web api if available or just use `!mp` command to BanchoBot.

</details>

**Comment by peppy** [comment]
We have hopes of adding all the tooling required for running tournaments to the osu client. Just a matter of finding the time and slotting it in priority-wise.
